[BoundsSafety] Fix diagnostics test failures

hnrklssn opened the pull request:
In fceb9cecdf6264eb773ee826b72a51a9ec68ec74 (https://github.com/llvm/llvm-project/pull/134769) upstream changed the diagnostic engine to always wrap expressions in single quotes. We already had manual wrapping for most of those, so the manual single quotes needed to be removed. In other placed the test cases needed to be updated to reflect the new output.

rdar://149359009

hnrklssn committed:
[BoundsSafety] Fix diagnostics test failures

In fceb9cecdf6264eb773ee826b72a51a9ec68ec74 upstream changed the
diagnostic engine to always wrap expressions in single quotes. We
already had manual wrapping for most of those, so the manual single
quotes needed to be removed. In other placed the test cases needed to be
updated to reflect the new output.

rdar://149359009

hnrklssn commented:
@swift-ci please test llvm

delcypher commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -13399,7 +13399,7 @@ def err_bounds_safety_terminated_by_wrong_initializer_kind : Error<
 def err_bounds_safety_terminated_by_terminator_mismatch : Error<
   "array '%0' with '__terminated_by' attribute is initialized with an "
-  "incorrect terminator (expected: %1; got %2)">;
+  "incorrect terminator (expected: %quoted1; got %quoted2)">;
```
This seems new. Looking at the implementation it looks like this just adds single quotes around the argument. Presumably this is related to https://github.com/llvm/llvm-project/pull/134769#issuecomment-2822234053 and now we have double single quotes in the outputted diagnostics?

delcypher commented on clang/test/BoundsSafety/Sema/terminated-by-array-init.c:
```diff
@@ -37,39 +37,39 @@ void explicit_const_init(void) {
+  // expected-error@+5{{array 'qux2.foo.a' with '__terminated_by' attribute is initialized with an incorrect terminator (expected: '0'; got '1')}}
+  // expected-error@+4{{array 'qux2.foo.b' with '__terminated_by' attribute is initialized with an incorrect terminator (expected: '42'; got '2')}}
+  // expected-error@+3{{array 'qux2.bar.a' with '__terminated_by' attribute is initialized with an incorrect terminator (expected: '0'; got '3')}}
+  // expected-error@+2{{array 'qux2.baz.a' with '__terminated_by' attribute is initialized with an incorrect terminator (expected: ''X''; got '0')}}
```
Ok so the character literal is double single quoted, and the integer literal is single quoted. It looks a little weird but I guess this is consistent with what upstream currently has.

delcypher reviewed (approved):
I think the double single quotes looks a little weird but I think we should land the test fixes first and debate this separately. These tests being broken is blocking me so I'd rather we just landed this.

delcypher commented:
For macOS

```
********************
Failed Tests (7):
  Clang :: BoundsSafety/CodeGen/bounds-attributed-in-return-null-system-header-O2.c
  Clang :: BoundsSafety/CodeGen/nested-struct-member-count-O2.c
  Clang :: BoundsSafety/CodeGen/terminated-by-to-indexable-trivial-O2.c
  Clang :: BoundsSafety/FixIt/null_terminated_bidi_indexable_conv.c
  Clang :: BoundsSafety/Sema/terminated-by-attr.c
  Clang :: SemaCXX/warn-unsafe-buffer-usage-count-attributed-pointer-argument.cpp
  Clang :: SemaCXX/warn-unsafe-buffer-usage-libc-functions-interop.cpp
```

@hnrklssn Are these failures unrelated?

hnrklssn committed:
[BoundsSafety] Fix failing codegen tests (NFC)

Upstream now folds (gep inbounds null, X) -> null. This updates our O2
tests to reflect the new output.
Caused by 5a993558c5fa52c605d984e0effdc1cd3b452476
(https://github.com/llvm/llvm-project/pull/130742).

hnrklssn committed:
[BoundsSafety] Fix accidentally changed diagnostic

hnrklssn commented:
> For macOS
> 
> ```
> ********************
> Failed Tests (7):
>   Clang :: BoundsSafety/CodeGen/bounds-attributed-in-return-null-system-header-O2.c
>   Clang :: BoundsSafety/CodeGen/nested-struct-member-count-O2.c
>   Clang :: BoundsSafety/CodeGen/terminated-by-to-indexable-trivial-O2.c
>   Clang :: BoundsSafety/FixIt/null_terminated_bidi_indexable_conv.c
>   Clang :: BoundsSafety/Sema/terminated-by-attr.c
>   Clang :: SemaCXX/warn-unsafe-buffer-usage-count-attributed-pointer-argument.cpp
>   Clang :: SemaCXX/warn-unsafe-buffer-usage-libc-functions-interop.cpp
> ```
> 
> @hnrklssn Are these failures unrelated?

These seem unrelated, the others should be fixed now. Merging to unblock.
```
Clang :: BoundsSafety/FixIt/null_terminated_bidi_indexable_conv.c
Clang :: BoundsSafety/Sema/terminated-by-attr.c
```
